README.txt
= WDDX

http://wddx.rubyforge.org/

== DESCRIPTION

Ruby API for the WDDX XML interchange format (see http://www.openwddx.org/) 

From http://www.openwddx.org/faq/:
 WDDX is an XML-based technology that enables the exchange of complex data between Web programming languages, 
 creating what some refer to as 'Web syndicate networks'. 
 WDDX consists of a language-independent representation of data based on an XML 1.0 DTD, and a set of 
 modules for a wide variety of languages that use WDDX. 
 WDDX can be used with HTTP, SMTP, POP, FTP and other Internet protocols that support transferring textual data.

WDDX Home: http://www.openwddx.org/

The Ruby WDDX Gem enables easy usage of WDDX in Ruby.

== Author

Copyright (C) 2005, 2006, 2007, 2008 by Stefan Saasen <s-NOSPAM-REMOVE@juretta.com> - http://juretta.com/

== Installation

Just run <tt>[sudo] gem install wddx</tt> to install the WDDX Gem.


== SYNOPSIS

=== Deserialization
You can very easily deserialize WDDX data from a file or from a string.

Simple usage example:
 
 require 'rubygems'
 require 'wddx'        
 # data.xml contains <?xml version="1.0">\
 # <wddxPacket version='1.0'><header><comment>A comment</comment></header><data><string>Klaus Tester</string></data></wddxPacket> 
 wddx = WDDX.load(File.open("data.xml"))
 puts wddx.comment # "A comment"
 puts wddx.data.class # String
 puts wddx.data # "Klaus Tester"
 
WDDX.load accepts a string with xml content, a file object or any object that provides a +read+ method.
 packet =  WDDX.load(open("http://wddx.rubyforge.org/wddx.xml"))
or
 packet =  WDDX.load(File.open("data.xml")) 
or
 packet =  WDDX.load(open("data.xml"))
or
 xml = "<wddxPacket version='1.0'><header/><data><number>123.667</number></data></wddxPacket>"
 packet = WDDX.load(xml)
 puts w.data # => 123.667
 
 xml = <<-EOF 
 <wddxPacket version='1.0'>
 <header/>
 <data>
     <struct>
         <var name='aNull'>
             <null/>
         </var>
         <var name='aString'>
             <string>a string</string>
         </var>
     </struct>
    </data>
  </wddxPacket> 
 EOF
 w = WDDX.load(xml)
 p w.data # => {"aString"=>"a string", "aNull"=>nil}
                                         
If the root element is a +struct+ you can use a shortcut version:
 
 w.data["aString"] # => "a string"

can be written as:

 w.aString # => "a string"


=== Serialization
The core Ruby Classes +Symbol+, +String+, +Numeric+, +true+, +false+, +nil+, +Hash+, +Array+ and +Time+ can be serialized by calling
+to_wddx+.

 require 'wddx'
 "Stefan Saasen".to_wddx        # => "<wddxPacket version='1.0'><header/><data>\
                                #     <string>Stefan Saasen</string></data></wddxPacket>"

 123.667.to_wddx                # => "<wddxPacket version='1.0'><header/><data>\
                                #     <number>123.667</number></data></wddxPacket>"

 Math::PI.to_wddx               # => "<wddxPacket version='1.0'><header/><data>\
                                #   <number>3.14159265358979</number></data></wddxPacket>"

 [1000, "Klaus Tester"].to_wddx # => "<wddxPacket version='1.0'><header/><data>\
                                #   <array length='2'><number>1000</number>\
                                #   <string>Klaus Tester</string></array></data></wddxPacket>"

=== WDDX.dump
You can use WDDX.dump to serialize Ruby objects.
 WDDX.dump("Hallo Welt") # => "<wddxPacket version='1.0'><header/><data><string>Hallo Welt</string></data></wddxPacket>"


Custom Ruby classes can be serialized by adding a +to_wddx_properties+ method (in the style of the YAML library) and calling WDDX.dump.

 class SerializeFromOutside
   def initialize(a,b,c)
     @a, @b, @c = a, b, c
   end
   
   def to_wddx_properties
     ["@a", "@c"]
   end  
 end 
 obj = SerializeFromOutside.new("This is a", "This is b", 123) 
 WDDX.dump(obj) # => "<wddxPacket version='1.0'><header/><data><struct><var name='a'>\
                # <string>This is a</string></var><var name='c'><number>123</number>\
                # </var></struct></data></wddxPacket>"

It is possible to include the WDDX module to get a behaviour similiar to the core Ruby classes.

 class MyObject
   include WDDX
   attr_accessor :name, :value, :price

   def to_wddx_properties
     ["@name", :custom_price]
   end               

   def custom_price
     @price * 1.05
   end
 end

 obj = MyObject.new("Stefan Saasen", 120)
 puts ob.to_wddx
 # => "<wddxPacket version='1.0'><header/><data><struct><var name='name'>\
 # <string>Stefan Saasen</string></var><var name='custom_price'><number>126.0</number>\
 # </var></struct></data></wddxPacket>"     
                                    
== WDDX Data types

WDDX defines some data types which can not be mapped to native ruby classes. WDDX::Binary represents a binary object (which is in fact a BASE64 encoded String). WDDX::RecordSet represents a RecordSet with data rows an column meta information.

== Type mapping
The following table shows the data type mapping WDDX <=> Ruby.
 WDDX Type      Ruby Type
 ---------      ---------------------
 String         String
 Number         Numeric
 Boolean        TrueClass, FalseClass
 Datetime       Time
 Null           nil
 Binary         WDDX::Binary
 Array          Array
 Struct         Hash
 Recordset      WDDX::RecordSet


== Rails Plugin

You can add the wddx gem to your rails application to be able to serialize ActiveRecord objects to WDDX.

In config/environment.rb add:
 require 'wddx'

Example:
 
 ...
 a_ar_obj.to_wddx # => <wddxPacket...


== See

Homepage:: http://rubyforge.org/projects/wddx/
Blog:: http://juretta.com/
Bugtracker:: http://rubyforge.org/tracker/?group_id=2715
WDDX:: http://www.openwddx.org/

== LICENSE:

(The MIT License)

Copyright (c) 2005, 2006, 2007, 2008 by Stefan Saasen <s-NOSPAM-REMOVE@juretta.com> - http://juretta.com/

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.4.1 / 2008-04-31
* Minor code cleanup.
* WDDX library now works in Ruby 1.8 and 1.9

=== 0.4.0 / 2008-03-04
* Added wddx binary. Pretty prints a wddx data structure.

  Usage: 
	wddx FILE|URL
  OR
    wddx < wddx.xml
  OR
    cat /path/to/wddx.xml | wddx

=== 0.3.3 / 2008-03-02
* Fixed test issues when running the test in a different timezone.

=== 0.3.2 / 2007-08-22
* Bugfix release: Fixes bug 13288.

=== 0.3.1 / 2007-04-19
* Bugfix release: Fixes bugs 10148 and 10149 reported by Rhett Sutphin
* Patch [#10152] kindly provided by Rhett Sutphin!

=== 0.3.0 / 2007-04-07
* New example wddx_serialize.rb added (see examples/ subdirectory).
* WDDX::Struct now includes Enumerable
* Documentation improved
* Symbol now responds_to?(:to_wddx)

=== 0.2.1 / 2006-12-20
Bugfix Release
* Fixed TcWddxComplexTest 

=== 0.2.0 / 2006-12-19
Added Rails support

Rails ActiveRecord::Base objects can be serialised to WDDX-XML using the +to_wddx+ method.

=== 0.1.0 / 2006-12-18
Initial release


lib/wddx.rb
#--
# $Id: wddx.rb 115 2007-08-22 21:11:31Z stefan $
#
# Copyright (C) 2005, 2006 by Stefan Saasen <s@juretta.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#++                                      
require 'open-uri'
#$:.unshift(File.dirname(__FILE__) + "/wddx") unless $:.include?(File.dirname(__FILE__) + "/wddx") || $:.include?(File.expand_path(File.dirname(__FILE__) + "/wddx"))

Dir[File.join(File.dirname(__FILE__), 'wddx/**/*.rb')].sort.each { |lib| require lib }           

# == Usage
#:include:README.txt
module WDDX
  # class methods
  class << self
    # Reads WDDX-XML from obj. Obj can be a String or every Object that responds to a "read" method (File, open-uri...)
    #  require 'open-uri'
    #  packet = WDDX.load(open("http://wddx.rubyforge.org/wddx.xml"))
    def open(obj, &block)
      if obj.respond_to?(:read)
        return deserialize(obj.read)
      end
      deserialize(obj)
    end
    alias :load :open
    
    # Reads WDDX-XML from the file +filename+.
    def from_file(filename, &block)
      deserialize(open(filename))
    end          
    alias :file :from_file
    
    # Deserialize WDDX-XML to a WDDX::WddxPacket
    #  p = WDDX.deserialize(xml_string)
    #  p.data # => String, Number, Time...
    def deserialize(xml_string)       
      Deserializer::Deserializer.new(xml_string).wddx_packet
    end
    
    # The method dump tries to serialize arbitrary Ruby objects.
    #  print WDDX.dump("Hello world") # => "<wddxPacket version='1.0'><header/><data><string>Hello world</string></data></wddxPacket>" 
    # 
    # You can add a +to_wddx_properties+ method to your class that returns an array of fields an method names (as symbol) that will be
    # serialized as WDDX XML.
    #
    #  class MyObj
    #    def initilize(name, price)
    #      @name, @price = name, price
    #    end
    #    def custom_price; @price * 1.05; end 
    #    
    #    def to_wddx_properties
    #      ["@name", :custom_price]
    #    end
    #  end
    #
    #  WDDX.dump(MyObj.new("Klaus Tester", 123))
    def dump(var)
      s = WDDX::Serializer::Serializer.new
      s.add_var var
      s.to_xml
    end
      
  end # class << self
  
  # The WDDX module can be mixed into arbitrary classes.
  # The method +wddx_serialize+ serializes the instance variables
  # or the instance variables and methods output defined in +to_wddx_properties+
  def wddx_serialize #:nodoc:
    hash = {}
    if self.respond_to?(:to_wddx_properties)
      self.send(:to_wddx_properties).each do |prop|
        if prop.respond_to?(:id2name)
          hash[prop.id2name] = self.send(prop)
        elsif prop =~ /^@[^@]/
          hash[prop.gsub(/@/, "")] = instance_variable_get(prop)
        end
      end
    else
      hash = self.instance_variables.sort.inject({}) {|h, var| h[var.to_s.gsub(/@/, "")] = instance_variable_get(var); h}
    end
    hash.wddx_serialize
  end
  
  # The WDDX module can be mixed into arbitrary classes. 
  # You can call +to_wddx+ to get a WDDX-XML representation of your object.
  #
  #  class MyTest
  #    include WDDX
  #    def initialize(a,b)
  #      @a, @b = a, b
  #    end
  #  end
  #  
  #  t = MyTest.new("Hello world", Time.now)
  #  print t.to_wddx # => "<wddxPacket ..."
  def to_wddx
    s = WDDX::Serializer::Serializer.new 
    s.add_var self
    s.to_xml
  end
  
  # = Binary
  # The binary datatype represents strings (blobs) of binary data. The WDDX 
  # DTD allows for multiple encodings of binary data. In this version only 
  # MIME style base64 encoding is supported by the specification. Optionally,
  # the length of the encoded binary object can be provided as a hint to
  # WDDX deserializers. It can be used to validate the length of the binary
  # object after decoding. It can also be used for efficient allocation of
  # memory during the decoding process.
  class Binary < WddxData
    attr_accessor :bin_data                                          
    
    # Initialize with the raw binary data (to be encoded with Base64)
    def initialize(bin_data = nil)
      @bin_data = bin_data
      @_data = nil
    end                   

    # set binary data
    def encoded_data=(arg)
       @bin_data = arg.unpack( 'm' )[0]
    end
    
    # length of the raw binary data in bytes
    def length
      @bin_data.length
    end       
    alias :size :length
    
    # Returns an Base64 encoded string
    def encode
      return @_data unless @_data.nil?
      @_data = [@bin_data].pack( 'm' ).chomp if @bin_data   
      @_data
    end
    
    # Returns the WDDX XML respresentation
    def wddx_serialize # :nodoc:
      "<binary length='#{self.length}'>#{self.encode}</binary>"
    end
    
    alias :data :bin_data
  end
  
end 

# Add "to_wddx"
[WDDX::Struct, WDDX::Binary, WDDX::RecordSet].each {|clazz| clazz.class_eval { include WDDX::Core } }
